src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -38,8 +41,50 @@ Q_LOGGING_CATEGORY(lcCfApiWrapper, "nextcloud.sync.vfs.cfapi.wrapper", QtInfoMsg
       FIELD_SIZE( CF_OPERATION_PARAMETERS, field ) )
 
 namespace {
+
+Q_GLOBAL_STATIC(std::mutex, activeRequestsMutex);
```

Comment:
I am not convinced that has to be there instead of in `OCC::VfsCfApi`. All call contexts allow to get a pointer to OCC::VfsCfApi and should allow to keep state tracking there.
You would need to extend the API and we could have a complete view of HydrationJob instances (deleted one, not deleted ones, ...)

Reply:
When we do this, we would need to add a method like `hydrationRequestFinished()` to `OCC::VfsCfApi` that allows `OCC::VfsCfApi` to free the hydration jobs. `OCC::VfsCfApi` inherits from `OCC::Vfs`. The methods that request hydration are part of the of `OCC:Vfs`. We could either add the `hydrationRequestFinished()` method to `OCC::Vfs` or only to `OCC::VfsCfApi`. I would find it awkward to add the `hydrationRequestFinished()` method only to `OCC::VfsCfApi`. On the other hand, if we add `hydrationRequestFinished()` to the `OCC::Vfs` it is awkward too, because `OCC::Vfs` already provides signals like `onFinishedHydration()`. I know we could keep a list of deleted hydration jobs, and cancelled ones, but at some point we need to clear them up. Otherwise the list will grow to infinity. That's why I think we need such a `hydrationRequestFinished()` method. I don't like this approach, because I have the feeling that cfapiwrapper and `OCC::VfsCfApi` become tightly coupled when doing this change. Maybe it would be the best, if cfapiwrapper becomes part of `OCC::VfsCfApi`, but this may be problematic because of the multi threading. Another problem is that these`HydrationJob` objects are not really doing the hydration. The more like download only jobs. Big part of the hydration is also done in cfapiwrapper.

What do you think?